### PR TITLE
Output XDMF that conforms to spec

### LINKED ISSUE #819: XDMF output is non-compliant with spec
The XDMF output is invalid: it does not validate against the XDMF DTD specification (https://gitlab.kitware.com/xdmf/xdmf/-/blob/master/Xdmf.dtd).

This can be verified by attempting to parse an XDMF file with `lxml` (with `Xdmf.dtd` present in the same directory):
```
In [1]: import lxml

In [2]: from lxml import etree

In [3]: parser = etree.XMLParser(dtd_validation=True)

In [4]: tree = etree.parse("./parthenon.prim.00000.phdf.xdmf", parser)
Traceback (most recent call last):
  File ~/Library/Python/3.10/lib/python/site-packages/IPython/core/interactiveshell.py:3378 in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  Cell In [4], line 1
    tree = etree.parse("./advection.out0.00000.phdf.xdmf", parser)
  File src/lxml/etree.pyx:3538 in lxml.etree.parse
  File src/lxml/parser.pxi:1876 in lxml.etree._parseDocument
  File src/lxml/parser.pxi:1902 in lxml.etree._parseDocumentFromURL
  File src/lxml/parser.pxi:1805 in lxml.etree._parseDocFromFile
  File src/lxml/parser.pxi:1177 in lxml.etree._BaseParser._parseDocFromFile
  File src/lxml/parser.pxi:615 in lxml.etree._ParserContext._handleParseResultDoc
  File src/lxml/parser.pxi:725 in lxml.etree._handleParseResult
  File src/lxml/parser.pxi:654 in lxml.etree._raiseParseError
  File ./advection.out0.00000.phdf.xdmf:10
    <Topology Type="3DRectMesh" NumberOfElements="2 17 17"/>
                                                         ^
XMLSyntaxError: No declaration for attribute Type of element Topology, line 10, column 61
```

The issue is that XDMF specifies that:
* `Type` should actually be `TopologyType` (https://gitlab.kitware.com/xdmf/xdmf/-/blob/master/Xdmf.dtd#L34)
* `NumberOfElements` should actually be `Dimensions` (https://gitlab.kitware.com/xdmf/xdmf/-/blob/master/Xdmf.dtd#L35).

### PR DESCRIPTION
**Pull request by BenWibking**
## PR Summary

This changes the XDMF output to conform to the current XDMF specification (available here: https://gitlab.kitware.com/xdmf/xdmf/blob/master/Xdmf.dtd).

To fix the data format issues, this also incorporates the changes proposed in https://github.com/parthenon-hpc-lab/parthenon/pull/796.

Fixes https://github.com/parthenon-hpc-lab/parthenon/issues/819.

## PR Checklist

- [x] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Code is formatted
- [x] Changes are summarized in CHANGELOG.md
- [ ] CI has been triggered on [Darwin](https://re-git.lanl.gov/eap-oss/parthenon/-/pipelines) for performance regression tests.
- [ ] Docs build
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Comment by BenWibking** [comment]
I've tested this on the outputs from the advection example with Paraview 5.11.0 and it works fine. With the same output files, VisIt 3.22 still crashes.

**Comment by BenWibking** [comment]
The crashes are unrelated, and are fixed by https://github.com/parthenon-hpc-lab/parthenon/pull/796.

**Review by Yurlungur — APPROVED** [review]
So long as Visit's happy, I'm happy. I see some chatter on the Matrix channel, so I defer to you and @brryan about how to proceed.

**Comment by BenWibking** [comment]
Visit works, but Paraview is still unhappy (emits lots of identical warnings).

**Comment by Yurlungur** [comment]
Ah ok.

**Comment by BenWibking** [comment]
Ok, it works for me now in both VisIt and Paraview without any warnings. @Yurlungur @brryan does this look good to you?

**Review by Yurlungur — APPROVED** [review]
I think this makes sense.

What output did you try this with? Before we merge I want to make sure visit/paraview are happy with both scalar and vector output and with 1 and many blocks. But assuming that's the case, I approve.

**Comment by BenWibking** [comment]
I tried this with both the advection example outputs and a problem in AthenaPK. I think (?) that covers both scalar and vector output, but I didn't test a single block simulation.

**Comment by Yurlungur** [comment]
> I tried this with both the advection example outputs and a problem in AthenaPK. I think (?) that covers both scalar and vector output, but I didn't test a single block simulation.

Good enough for me.

File: src/outputs/parthenon_hdf5.cpp [hunk 1]
```diff
@@ -264,11 +264,12 @@ static std::string stringXdmfArrayRef(const std::string &prefix,
                                       const std::string &hdfPath,
                                       const std::string &label, const hsize_t *dims,
                                       const int &ndims, const std::string &theType,
-                                      const int &precision) {
+                                      const int &precision, bool isVector) {
```
Comment (pgrete) [inline src/outputs/parthenon_hdf5.cpp]: Why this change?
`isVector` does not seem to be used.

File: src/outputs/parthenon_hdf5.cpp [hunk 2]
```diff
@@ -264,11 +264,12 @@ static std::string stringXdmfArrayRef(const std::string &prefix,
...
   std::string mystr =
-      prefix + R"(<DataItem Format="HDF" Dimensions=")" + std::to_string(dims[0]);
-  for (int i = 1; i < ndims; i++)
+      prefix + R"(<DataItem Format="HDF" Dimensions=")"; // + std::to_string(dims[0]);
```
Comment (pgrete) [inline src/outputs/parthenon_hdf5.cpp]:
```suggestion
      prefix + R"(<DataItem Format="HDF" Dimensions=")";
```

File: src/outputs/parthenon_hdf5.cpp [hunk 3]
```diff
@@ -312,19 +315,17 @@ static void writeXdmfSlabVariableRef(std::ofstream &fid, const std::string &name
     }
     fid << ">" << std::endl;
     fid << prefix << "  "
-        << R"(<DataItem ItemType="HyperSlab" Dimensions=")" << dims321 << " "
-        << vector_size << R"(">)" << std::endl;
-    // TODO(JL) 3 and 5 are literals here, careful if they change.
-    // "3" rows for START, STRIDE, and COUNT for each slab with "5" (H5_NDIM) entries.
-    // START: iblock variable(_component)  0   0   0
-    // STRIDE: 1               1           1   1   1
-    // COUNT:  1           vector_size    nx3 nx2 nx1
```
Comment (pgrete) [inline src/outputs/parthenon_hdf5.cpp:321]: I think the comment is still valid.

File: src/outputs/parthenon_hdf5.cpp [hunk 4]
```diff
@@ -312,19 +315,17 @@ static void writeXdmfSlabVariableRef(std::ofstream &fid, const std::string &name
...
+        << R"(<DataItem ItemType="HyperSlab" Dimensions=")";
+    for (int i = 2; i < ndims; i++) {
+      fid << dims[i] << " ";
+    }
```
Comment (pgrete) [inline src/outputs/parthenon_hdf5.cpp]: Why this change (compared to using `dims321`)?
I think we should remain consistent, i.e., either switch to a loop everywhere and get rid of `dims321` or use `dims321` wherever possible (which should make future changes easier as all places can be identified by the same pattern)

File: src/outputs/parthenon_hdf5.cpp [hunk 5]
```diff
@@ -406,15 +405,30 @@ void genXDMF(std::string hdfFile, Mesh *pm, SimTime *tm, int nx1, int nx2, int n
...
+        dims[vinfo.tensor_rank + 1] = nx3;
+        dims[vinfo.tensor_rank + 2] = nx2;
+        dims[vinfo.tensor_rank + 3] = nx1;
+      } else {
+        continue;
+      }
+
       const int vlen = vinfo.vlen;
-      dims[1] = vlen;
+      // dims[1] = vlen;
```
Comment (pgrete) [inline src/outputs/parthenon_hdf5.cpp]:
```suggestion
```

File: src/outputs/parthenon_hdf5.cpp [hunk 6]
```diff
@@ -406,15 +405,30 @@ void genXDMF(std::string hdfFile, Mesh *pm, SimTime *tm, int nx1, int nx2, int n
     xdmf << "      </Geometry>" << std::endl;
 
     // write graphics variables
-    dims[1] = 1;
-    dims[2] = nx3;
-    dims[3] = nx2;
-    dims[4] = nx1;
+    int ndim;
     for (const auto &vinfo : var_list) {
+      // TODO(BRR) just let vinfo provide this
```
Comment (pgrete) [inline src/outputs/parthenon_hdf5.cpp]: @brryan is that comment still valid (as `vinfo` is already used)?

**Review by pgrete — CHANGES_REQUESTED** [review]
Mostly requesting cleanups. I can also confirm that the patch works (though I didn't test non-vector variables).

**Comment by BenWibking** [comment]
@brryan Can you address @pgrete's comments above? I think these changes originally came from your PR. (Alternatively, you could merge my changes into your PR.)

File: src/outputs/parthenon_hdf5.cpp (hunk 1 above)
Comment (brryan) [reply src/outputs/parthenon_hdf5.cpp]: Yes this was a mistake, thanks for noticing this

File: src/outputs/parthenon_hdf5.cpp (hunk 3 above)
Comment (brryan) [reply src/outputs/parthenon_hdf5.cpp:321]: :+1:

File: src/outputs/parthenon_hdf5.cpp (hunk 4 above)
Comment (brryan) [reply src/outputs/parthenon_hdf5.cpp]: No reason now for this change, and in fact it was overly specialized to the 1D tensor dimension case -- reverted. I agree about `dims321` consistency.

File: src/outputs/parthenon_hdf5.cpp (hunk 6 above)
Comment (brryan) [reply src/outputs/parthenon_hdf5.cpp]: I had thought about storing `alldims` itself in `vinfo` but I don't think that's a meaningful change -- comment removed

**Comment by brryan** [comment]
> @brryan Can you address @pgrete's comments above? I think these changes originally came from your PR. (Alternatively, you could merge my changes into your PR.)

Thanks for pinging me on this @BenWibking, yes I am responsible for these issues :). I made the fixes in response to @pgrete's comments in #796. While looking at this I noticed that scalar variable output wasn't working with XDMF, e.g. the `in_or_out` variable from the `calculate_pi` example (I think both `advection` and any AthenaPK problem only produce vector-type output (AthenaPK has one big state vector for all variables as I understand it)). I modified `writeXdmfSlabVariableRef` to work for the scalar case as well. Similar modifications would need to be made for 2- and 3-tensor dimension variables but since this work is already spread across two PRs and currently only I (?) use higher-tensor-dimension variables I'll open an issue instead. 

Since the changes I made weren't entirely trivial could I trouble you to merge in my latest `brryan/xdmf_fix` and check that paraview is still happy (I only checked `burgers`, `advection`, and `calculate_pi` in visit)? Then if everything works I think we can just merge this PR in.

**Comment by BenWibking** [comment]
> Since the changes I made weren't entirely trivial could I trouble you to merge in my latest `brryan/xdmf_fix` and check that paraview is still happy (I only checked `burgers`, `advection`, and `calculate_pi` in visit)? Then if everything works I think we can just merge this PR in.

I've checked that `burgers`, `advection` and `calculate_pi` all load into Paraview without any issues.

**Review by pgrete — APPROVED** [review]

**Comment by pgrete** [comment]
I removed the unused parameter, formatted the code, and added a changelog.
Enabling auto-merge now.

**Comment by brryan** [comment]
> > Since the changes I made weren't entirely trivial could I trouble you to merge in my latest `brryan/xdmf_fix` and check that paraview is still happy (I only checked `burgers`, `advection`, and `calculate_pi` in visit)? Then if everything works I think we can just merge this PR in.
> 
> I've checked that `burgers`, `advection` and `calculate_pi` all load into Paraview without any issues.

Thanks @BenWibking!

> I removed the unused parameter, formatted the code, and added a changelog. Enabling auto-merge now.

Thanks @pgrete! (sorry I missed the unused parameter)

**Comment by Yurlungur** [comment]
The Python/C++ formatting static check does not seem to be triggering. I tried enabling/disabling auto-merge to no effect. Is something wrong?

**Comment by Yurlungur** [comment]
Looks like the format check was skipped for some reason.

**Comment by Yurlungur** [comment]
I cloned the branch and ran formatting manually. It passes, so I'm going to force merge. @pgrete I think there maybe some extra step needed to get the formatting workflow to work on forks?

**Pull request merged**